src/libopensc/card-dnie.c

```diff
@@ -729,6 +729,62 @@ int dnie_match_card(struct sc_card *card)
 	LOG_FUNC_RETURN(card->ctx, result);
 }
 
+static int dnie_sm_free_wrapped_apdu(struct sc_card *card,
+		struct sc_apdu *plain, struct sc_apdu **sm_apdu)
+{
+	struct sc_context *ctx = card->ctx;
+	int rv = SC_SUCCESS;
+
+	LOG_FUNC_CALLED(ctx);
+	if (!sm_apdu)
+		LOG_FUNC_RETURN(ctx, SC_ERROR_INVALID_ARGUMENTS);
+	if (!(*sm_apdu))
+		LOG_FUNC_RETURN(ctx, SC_SUCCESS);
+
+	plain->resp = (*sm_apdu)->resp;
+	plain->resplen = (*sm_apdu)->resplen;
```

Comment:
If some memory for `plain->resp` has already been initialized this introduces a memory leak. If I'm not mistaken then `get_sm_apdu`/`free_sm_apdu` should only do the memory handling for the encrypted APDU. Allocating memory of the unencrypted response data is managed by the application (i.e. something like opensc-pkcs11 or opensc-explorer).